command/ca/certificate.go

```diff
@@ -443,3 +486,408 @@ func parseTimeDuration(ctx *cli.Context) (notBefore api.TimeDuration, notAfter a
+					return err
+				}
+				chValidated = true
+				break
+			}
+		}
+		if !chValidated {
+			if az.Wildcard {
+				return errors.Errorf("wildcard dnsnames (%s) require dns validation, "+
+					"which is currently not implemented in this client", ident)
+			}
+			return errors.Errorf("unable to validate any challenges for identifier: %s", ident)
+		}
+	}
+	return nil
+}
+
+func finalizeOrder(ac *ca.ACMEClient, o *acme.Order, csr *x509.CertificateRequest) (*acme.Order, error) {
+	var (
+		err              error
+		ro, fo           *acme.Order
+		isReady, isValid bool
+	)
+	ui.Printf("Waiting for Order to be 'ready' for finalization .")
+	for i := 9; i >= 0; i-- {
+		time.Sleep(1 * time.Second)
+		ui.Printf(".")
+		ro, err = ac.GetOrder(o.ID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error retrieving order %s", o.ID)
+		}
+		if ro.Status == "ready" {
+			isReady = true
+			ui.Printf(" done!\n")
+			break
+		}
+	}
+	if !isReady {
+		ui.Printf(" Error!\n\n")
+		return nil, errors.Errorf("Unable to validate order: %+v", ro)
+	}
+
+	ui.Printf("Finalizing Order .")
+	if err = ac.FinalizeOrder(o.Finalize, csr); err != nil {
+		return nil, errors.Wrapf(err, "error finalizing order")
+	}
+
+	for i := 9; i >= 0; i-- {
+		time.Sleep(1 * time.Second)
+		ui.Printf(".")
+		fo, err = ac.GetOrder(o.ID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error retrieving order %s", o.ID)
+		}
+		if fo.Status == "valid" {
+			isValid = true
+			ui.Printf(" done!\n")
+			break
+		}
+	}
+	if !isValid {
+		ui.Printf(" Error!\n\n")
+		return nil, errors.Errorf("Unable to finalize order: %+v", fo)
+	}
+
+	return fo, nil
+}
+
+func acmeFlow(ctx *cli.Context) error {
+	args := ctx.Args()
+	subject := args.Get(0)
+	crtFile, keyFile := args.Get(1), args.Get(2)
+
+	sans := ctx.StringSlice("san")
+
+	if len(sans) == 0 {
+		sans = []string{subject}
+	}
+	dnsNames, ips := splitSANs(sans)
+	if len(ips) > 0 {
+		return errors.New("IP Address SANs are not supported for ACME flow")
+	}
+
+	caURL := ctx.String("ca-url")
+	if len(caURL) == 0 {
+		return errs.RequiredFlag(ctx, "ca-url")
+	}
+	if !(strings.HasSuffix(caURL, "directory") || strings.HasSuffix(caURL, "dir")) {
+		caURL = caURL + "/acme/directory"
+	}
+
+	var idents []acme.Identifier
+	for _, dns := range dnsNames {
+		idents = append(idents, acme.Identifier{
+			Type:  "dns",
+			Value: dns,
+		})
+	}
+
+	var (
+		err          error
+		orderPayload []byte
+		clientOps    []ca.ClientOption
+	)
+	if strings.Contains(caURL, "letsencrypt") {
+		// LetsEncrypt does not support NotBefore and NotAfter attributes in orders.
+		if ctx.IsSet("not-before") || ctx.IsSet("not-after") {
+			return errors.New("LetsEncrypt public CA does not support NotBefore/NotAfter " +
+				"attributes for certificates. Instead, each certificate has a default lifetime of 3 months.")
+		}
+		// Use default transport for public CAs
+		clientOps = append(clientOps, ca.WithTransport(http.DefaultTransport))
```

Comment:
This won't work with staging, we can use https://golang.org/pkg/crypto/x509/#SystemCertPool + the `--root` flag. We might be adding ours if defined in `defaults.json` or if we look for it like we do in other commands, but it should be ok.

Reply:
When you say this won't work for staging, what does that mean? If you're talking about letsencrypt staging, then it still works. I admit the logic won't work for other public facing ACME CAs whose url does not include "letsencrypt".

Reply:
I mean let's encrypt staging environment, and yes, it works because the directory and endpoints are served using a publicly valid certificate, but the certs that they serve are signed by an internal certificate. 

No issue, I thought they were also using their internal staging certificate to sign the endpoints.